### ISSUE TITLE
If logged-in user is unauthorized to do something he is being logged-out automatically

### ISSUE THREAD

**Issue by KrzysztofMadejski** [ckan/ckan#2381]
Concerns: CKAN 2.3

I guess it's because how NotAuthorized is handled by repoze [because it's handled correctly :)]. 

I don't see in CKAN distinction between Unauthorized (http 401 meaning `performing this action requires authentication and user is not authenticated`) and Forbidden (http 403 meaning `user is logged in but don't have rights to do given action; don't try to authenticate again`). I think it's quite crucial.

How to do this? One could think of sth like below:

``` python
        try:
            # Do not query for the group datasets when dictizing, as they will
            # be ignored and get requested on the controller anyway
            data_dict['include_datasets'] = False
            c.group_dict = self._action('group_show')(context, data_dict)
            c.group = context['group']
        except NotFound:
            abort(404, _('Group not found'))
        except NotAuthorized:
            abort(401, _('Unauthorized to read group %s') % id)
        except Forbidden:
            abort(403, _('Unauthorized to read group %s') % id)
```

But those exceptions already carry a message. No need to override it if we're not changing it. It would be better if one could write simply:

``` python
        # Do not query for the group datasets when dictizing, as they will
            # be ignored and get requested on the controller anyway
            data_dict['include_datasets'] = False
            c.group_dict = self._action('group_show')(context, data_dict)
            c.group = context['group']
```

and some wrapper around controllers would try-catch and abort selectively:

``` python
try:
    perform_action()
except NotFound as e:
    abort(404, e.msg)
except NotAuthorized as e:
    abort(401, e.msg)
except Forbidden as e:
    abort(403, e.msg)
```

What do you think?

**Comment by amercader** [comment]
This sounds like a great proposal @KrzysztofMadejski, and should be straight-forward (if a bit tedious) to change the exception handling on the controllers as you suggest. Do you have time to send a PR?

**Comment by KrzysztofMadejski** [comment]
Sadly I don't have enough time to prepare this PR. 

BTW: Do you have an idea how one could implement this try-catch wrapper wround controllers? Looking at stack-frames control jumps out right to pylons. I see two options, but I'm a ckan-pylons newbie so correct me if I'm wrong:

1. Write a middleware inserted somewhere [question is where it should be positioned among other middleware classes]
2. Override BaseController `__getattr__` to wrap every method in try-catch.

**Comment by Aung-Myint-Thein** [comment]
We have found a way to fix this. Not sure if it is the right way but it works :D 

We found that the issue is in abort function from ckan.lib.base.py which calls _abort from pylons.controllers.util because it is the ultimate place to go after showing all the message. So, we tried to modify abort function from ckan.lib.base.py in our extension. 

```
def abort(status_code=None, detail='', headers=None, comment=None):
    '''Abort the current request immediately by returning an HTTP exception.

    This is a wrapper for :py:func:`pylons.controllers.util.abort` that adds
    some CKAN custom behavior, including allowing
    :py:class:`~ckan.plugins.interfaces.IAuthenticator` plugins to alter the
    abort response, and showing flash messages in the web interface.

    '''
    if status_code == 401:
        # Allow IAuthenticator plugins to alter the abort
        for item in p.PluginImplementations(p.IAuthenticator):
            result = item.abort(status_code, detail, headers, comment)
            (status_code, detail, headers, comment) = result

    if detail and status_code != 503:
        h.flash_error(detail.encode('utf8'))
    # redirect to home page but this can be modified with a parameter too.
    h.redirect_to(controller='organization', action='index')
```

Comment and suggestions are welcome :D

**Comment by gleb-rudenko** [comment]
@amercader, is it still topical?

**Comment by wardi** [comment]
should have been fixed by #2846
